Develop

CezaryZal opened the pull request:
@Tomek
Add exceptions handling and corrected all errors indicated.
More information on the changes can be found in the following commits

CezaryZal committed:
Create `TokenBuilder` class for authentication client

Create the `SecurityConstants` to creating the token by login and password;
Create the `configurationData.properties`

CezaryZal committed:
Code reduction for better readability use service and controllers

Change of return type, from Optional to entity object;
Handling the Optional was add to `UserServiceImp`;
Renamed `ApiService` class to `LoginService`

CezaryZal committed:
Update `validator` package

The validation function has been divided into the new classes;
Create `PasswordFormValidator` and `LoginFormValidator` classes, which extend new created the `FormValidator` class;
The new `UserLoginValidatorService` class supports functions in the validator package

CezaryZal committed:
Add `PasswordComparator` class

The new class connection with `LoginService`

CezaryZal committed:
Remove `User` Object `foundUser` filed

CezaryZal committed:
Create exceptions handling

Create the `exception` package to store exception classes in it:
`EmptyObjectException`, `IncorrectInput`, `ToShortStringException`,
`IncorrectLoginOrPasswordSecurityException`;
Add the `ResponseEntity` class to all controller;
Add throwing exceptions to all class filters;

s235jr commented on src/main/java/com/CezaryZal/controller/LoginController.java:
```diff
@@ -1,24 +1,24 @@
 
-    private ApiService apiService;
+    private LoginService loginService;
 
```
@Autowired?

CezaryZal replied:
Ok, dobra praktyka

s235jr commented on src/main/java/com/CezaryZal/exceptions/IncorrectInput.java:
```diff
@@ -0,0 +1,8 @@
+package com.CezaryZal.exceptions;
+
+public class IncorrectInput extends RuntimeException {
```
jak to exception to nazwa powinna sie konczync na ...Exception

CezaryZal replied:
Ok

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -0,0 +1,64 @@
+        this.passwordComparator = passwordComparator;
+    }
+
+    public ResponseEntity<String> getTokenByUserLogin(UserLogin inputUserLogin) {
```
To nie powinno zwracać ResponseEntity a nieowrappowany  obiekt. ResponseEntity powinno się pojawić dopiero w warstwie controllera.

CezaryZal replied:
poprawione

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -0,0 +1,64 @@
+        String massage = "";
+        HttpStatus status = HttpStatus.OK;
+
+        try {
```
Spring dostarcza  mechanizm do obsługi exceptionów https://www.baeldung.com/exception-handling-for-rest-with-spring , myśle że warto z niego skorzystać

CezaryZal replied:
Stworzyłem oddzielny controller do obsługi wyjątków

s235jr commented on src/main/java/com/CezaryZal/manager/db/service/UserService.java:
```diff
@@ -2,13 +2,14 @@
-    Optional<User> findByLoginName(String loginName);
+    User findByLoginName(String loginName) throws NameNotFoundException;
 
     Iterable<User> findAll();
```
Poszukaj sobie metody, która zwróci Ci tutaj List<User> zamiast Iterable, z Listy będzie Ci łatwiej skorzystać niż z Iterable

CezaryZal replied:
Po za `UserController` nigdzie nie używam Iterable<T>.
Zmienię dla porządku

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/UserLoginValidatorService.java:
```diff
@@ -0,0 +1,37 @@
+
+    private LoginFormValidator loginFormValidator;
+    private PasswordFormValidator passwordFormValidator;
+
```
@Autowired

CezaryZal replied:
dobra praktyka, zmienione

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/UserLoginValidatorService.java:
```diff
@@ -0,0 +1,37 @@
+
+    private void throwIsEmptyUserLogin(UserLogin userLogin) {
+        if (loginFormValidator.isEmpty(userLogin)) {
+            throw new NullPointerException("User login is null");
```
Stworzyłbym tu włąsny exception

CezaryZal replied:
Ok, bardziej sprecyzuje

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/UserLoginValidatorService.java:
```diff
@@ -0,0 +1,37 @@
+            throwIsEmptyUserLogin(userLogin);
+            loginFormValidator.validLogin(userLogin.getLogin());
+            passwordFormValidator.validPassword(userLogin.getPassword());
+        } catch (RuntimeException exc){
```
Tak na pewno nie. Tworzysz z ogólnego exceptiona jakiś konkretny. Masz 100% pewności że ten RuntimeException będzie IncorrecIInputEXceptionem? Może sięzdarzyćże nie, to będzie ciężkie do zrozumienia z logów, co się tutaj zadziało

CezaryZal replied:
Ok, poprawiłem

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/FormValidator.java:
```diff
@@ -0,0 +1,20 @@
+@Service
+public class FormValidator {
+
+    private final int MIN_LENGTH = 3;
```
dodałbym jeszcze static

CezaryZal replied:
nawet jeżeli jest wykorzystywane tylko w tej klasie?

CezaryZal replied:
dobre praktyki, poprawione

CezaryZal committed:
Use good practice

Add annotation @Autowired on constructor;
Create `NullInputException` and swap it for the `NullPointerException`;
Rename `IncorrectInput`, add string `Exception`;
Add static in the `MIN_LENGTH` field

CezaryZal committed:
Update project to increase transparency code

In `UserController` and `UserService` classes change type, List from Iterable;
In `PasswordComparator` class change logic in void method;

CezaryZal committed:
Update handle exceptions, use Spring annotation

Create `RestResponseEntityExceptionHandler` class to handle all exception;
Remove `IncorrectInputException` class;
Generalization of the login exception using `IncorrectLoginOrPasswordSecurityException` class;
Rename description in validator classes

s235jr commented on src/main/java/com/CezaryZal/exceptions/RestResponseEntityExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+            EmptyObjectException.class,
+            ToShortStringException.class,
+            IncorrectLoginOrPasswordSecurityException.class})
+    protected ResponseEntity<Object> handleValidationLoginAndPasswordExceptions(RuntimeException ex, WebRequest request) {
+        return handleExceptionInternal(ex, ex.getMessage(), new HttpHeaders(), HttpStatus.BAD_REQUEST, request);
```
a nie lepiej zbudowaćsobie od raz u swoj włąsny response niż korzystać z jakieś handleexceptioninternal, który ma swoją logike?

CezaryZal replied:
Obecnie jest niepotrzebny racja, ma większą złożoność obliczeniową.
Chciałbym później rozbudować obiekt zwracany przy błedzie.
Dlatego też myślałem żeby na teraz stworzyc własny obiekt zwrcany

s235jr commented on src/main/java/com/CezaryZal/controller/LoginController.java:
```diff
@@ -19,8 +23,8 @@ public LoginController(LoginService loginService) {
     @PostMapping("/login")
-    public ResponseEntity<String> getTokenByUserLogin(@RequestBody UserLogin inputUserLogin){
-        return loginService.getTokenByUserLogin(inputUserLogin);
+    public ResponseEntity<String> getTokenByUserLogin(@RequestBody UserLogin inputUserLogin) throws NameNotFoundException, AccountNotFoundException {
```
czym się w sensie logicznym różni NameNotFoundException od AccountNotFoundException? Czy to są dwa odrębne wyjątki?

dodatkowo w innym miejscu masz metode: 

private void handleUserByActive(User foundUser) throws AccountNotFoundException {
        if (!foundUser.isActive()) {
            throw new AccountNotFoundException("The requested user has not been activated");
        }
    }

metod ma nazwe handle... sprawdza czy user jest aktywny, a jak nie jest aktywny to rzuca exception że user nieznaleziony

CezaryZal replied:
Exception NameNotFoundException został z poprzedniej logiki, nieusunołem rzucanego wyjątku.
Usunięte.
nazwę metode zmieniłem zgodnie z jej funkcją: `throwExceptionIfUserIsNotActive`

s235jr commented on src/main/java/com/CezaryZal/exceptions/EmptyObjectException.java:
```diff
@@ -1,5 +1,9 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+//@ResponseStatus(value = HttpStatus.BAD_REQUEST)
```
zbędne komentarze wywalamy

CezaryZal replied:
Tak, chciałem zapisać.
Przeniosę w notatnik i usunę z klasy

s235jr commented on src/main/java/com/CezaryZal/exceptions/RestResponseEntityExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+
+    @ExceptionHandler({IncorrectLoginOrPasswordSecurityException.class})
+    protected ResponseEntity<Object> handleLoginExceptions(RuntimeException ex, WebRequest request) {
+        return handleExceptionInternal(ex, ex.getMessage(), new HttpHeaders(), HttpStatus.PRECONDITION_FAILED, request);
```
412 po IncorrectLoginOrPasswordSecurityException ? nie pasuje mi tu raczej nic poza 401/403

CezaryZal replied:
Ok, zmienione na 403.

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -32,25 +30,14 @@ public LoginService(UserServiceImp userServiceImp, TokenBuilder tokenBuilder,
+    public String getTokenByUserLogin(UserLogin inputUserLogin) throws NameNotFoundException, AccountNotFoundException {
 
-        try {
             handleUserLogin(inputUserLogin);
```
handleUserLogin wywołuje userLoginValidator.isCorrectUserLogin(inputUserLogin); nie powinna zatem ta metoda nazywać się jakoś inaczej?

CezaryZal replied:
Racja, zmieniłem nazwy metod na bardziej prawdziwe/trafne

s235jr commented on src/main/java/com/CezaryZal/controller/UserControllerImp.java:
```diff
@@ -2,32 +2,37 @@
     @GetMapping("/name/{loginName}")
-    public Optional<User> findByLoginName(@PathVariable String loginName){
-        return userService.findByLoginName(loginName);
+    public ResponseEntity<User> findByLoginName(@PathVariable String loginName) throws NameNotFoundException {
```
zastanów się czy login w pathvariable jest dobrym pomysłem?
czy jakieś nietypowe znaki w loginie @#$%^&* nie będą problematyczne?

CezaryZal replied:
A czemu miały by być problematyczne skoro przesyłany jest łańcuch znaków, czy to znaki czy symbole dla aplikacji nie będę problemem

s235jr commented on src/main/java/com/CezaryZal/exceptions/EmptyObjectException.java:
```diff
@@ -0,0 +1,12 @@
+package com.CezaryZal.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
```
unused imports?

CezaryZal replied:
Ponieważ zakomentowałem @ResponseStatus, już tą adnotację usunąłem

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -0,0 +1,53 @@
+        userLoginValidator.isCorrectUserLogin(inputUserLogin);
+    }
+
+    private void handleUserByActive(User foundUser) throws AccountNotFoundException {
```
czy status że user jest aktywny/nieaktywny nie powinna być sprawdzana w filtrze przy każdym requescie?
Założmy że mam usera w Twoim serwisie, Ty jako admin deaktywujesz mi go, ale ja nadl mam dostęp do aplikacji tak długo jak długo będę zalogowany.

CezaryZal replied:
W tej aplikacji ma inne zastosowanie,

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -0,0 +1,53 @@
+    private void handleUserByActive(User foundUser) throws AccountNotFoundException {
+        if (!foundUser.isActive()) {
+            throw new AccountNotFoundException("The requested user has not been activated");
+        }
```
o Tym pisałem wyżej

CezaryZal replied:
Ok, już poprawiłem

s235jr commented on src/main/java/com/CezaryZal/manager/builder/TokenBuilder.java:
```diff
@@ -0,0 +1,32 @@
+
+    public String buildTokenByUser(User foundUser){
+        long currentTimeMillis = System.currentTimeMillis();
+        String secret = SecurityConstants.SECRET_KEY;
```
potrzebujemy to wynosić do zmiennej?

CezaryZal replied:
teoretycznie nie, ale jeśli chcemy precyzyjnie określić długość życia tokenu (co do milisecundy) to mozna zostawić.

CezaryZal replied:
Z dolnego problemu, zrobiłem jak zasugerowałeś

s235jr commented on src/main/java/com/CezaryZal/manager/builder/TokenBuilder.java:
```diff
@@ -0,0 +1,32 @@
+public class TokenBuilder {
+
+    public String buildTokenByUser(User foundUser){
+        long currentTimeMillis = System.currentTimeMillis();
```
czy to na pewno jest prawidłowo? czy currentTimeMillis nie dostanie wartości w momencie inicjowania beana i na zawsze będzie miał te samą wartość?

Czy token wygenerowany po czasie życia aplikacji dłuższym niż czas ważności tokenu będzie ważny?

CezaryZal replied:
Ooo to jest ważna uwaga.
Nie sprawdzałem dokładnie długości czasu tokenu więc patrząc na kod masz rację. 
Jest to czas inicjowania, wrzucę do buildera.

s235jr commented on src/main/java/com/CezaryZal/manager/builder/TokenBuilder.java:
```diff
@@ -0,0 +1,32 @@
+@Service
+public class TokenBuilder {
+
+    public String buildTokenByUser(User foundUser){
```
a co TokenBuildera interesuje co to za user i skąd on się wział, jemu to bez różnicy czy on jest znaleziony, czy się sam przypałętał, jego to wali (zmieniłbym nazwe z foundUser na user) :)

CezaryZal replied:
ok :P
Chciałem zprecyzować co do za user

s235jr commented on src/main/java/com/CezaryZal/manager/db/service/UserService.java:
```diff
@@ -2,15 +2,16 @@
 public interface UserService {
 
-    Optional<User> findById(Long index);
+    User findById(Long index) throws NameNotFoundException;
```
szukasz usera po ID jak nie znajdziesz to rzucasz exception że namenotfound - niespójne troche
może po prostu powinien być exception usernotfound

CezaryZal replied:
OK

s235jr commented on src/main/java/com/CezaryZal/manager/db/service/UserService.java:
```diff
@@ -2,15 +2,16 @@
-    Optional<User> findById(Long index);
+    User findById(Long index) throws NameNotFoundException;
 
-    Optional<User> findByLoginName(String loginName);
```
to samo, szukasz po loginie zwracasz że namenotfound, to niespójne

CezaryZal replied:
OK

s235jr commented on src/main/java/com/CezaryZal/manager/db/service/UserService.java:
```diff
@@ -2,15 +2,16 @@
+    User findByLoginName(String loginName) throws NameNotFoundException;
 
-    Iterable<User> findAll();
+    List<User> findAll();
```
OK, dużo lepiej, listką jest dużo łatwiej zarządzać

s235jr commented on src/main/java/com/CezaryZal/manager/db/service/UserServiceImp.java:
```diff
@@ -1,30 +1,35 @@
-        return userRepository.findById(index);
+    public User findById(Long index) {
+        return userRepository.findById(index)
+                .orElseThrow(() -> new IncorrectLoginOrPasswordSecurityException("Login or password incorrectly entered"));
```
bardziej rzuciłbym tutaj włąsnie jakiś wspomniany wyżej usernotfoundexception,
Sięgasz do repo po ID usera i jak go nie znajdziesz to rzucasz IncorrectLoginOrPasswordSecurityException. Wydaje mi się że to nie jest warstwa na ten exception.

CezaryZal replied:
W przypadku dostępu tylko administratrora można szczegółowiej okreslić dlaczego nie mamy dostępu do zasobów ale przy warstwie użytkownika wspominałeś że powinna być jeden ogólny wyjątek, kwestie bezpieczeństwa.

s235jr replied:
Ale czy wyjątek który rzucasz jest prawdziwy?
Szukasz użytkownika po ID i go nie znajdujesz.
Co do ma  tego ma login użytkownika?
Co do ma tego hasło użytkownika?

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -0,0 +1,53 @@
+        this.userServiceImp = userServiceImp;
+        this.tokenBuilder = tokenBuilder;
+        this.userLoginValidator = userLoginValidator;
+        this.passwordComparator = passwordComparator;
```
omówione w klasie PasswordCOmparator

CezaryZal replied:
Ok

s235jr commented on src/main/java/com/CezaryZal/manager/filters/comparator/PasswordComparator.java:
```diff
@@ -0,0 +1,14 @@
+package com.CezaryZal.manager.filters.comparator;
+
+import com.CezaryZal.exceptions.IncorrectLoginOrPasswordSecurityException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class PasswordComparator {
+
+    public void throwIsNotEqualsPassword(String passwordFromUserLogin, String userPasswordFromDb) {
+        if (!passwordFromUserLogin.equals(userPasswordFromDb)){
+            throw new IncorrectLoginOrPasswordSecurityException("Login or password incorrectly entered");
+        }
+    }
+}
```
Ta klasa raczej nie powinna istnieć, masz w springsecurity mechanizm do tego.

public AuthenticationResponse authenticate(AuthenticationRequest authenticationRequest) {
        UsernamePasswordAuthenticationToken authenticationToken =
                new UsernamePasswordAuthenticationToken(authenticationRequest.getUsername(), authenticationRequest.getPassword());
        Authentication authentication = authenticationManager.authenticate(authenticationToken);
        if (authentication.isAuthenticated()) {
            String jwt = jwtBuilder.build(authentication, AuthenticationConstants.JWT_EXPIRE_IN_MINUTES);
            RefreshToken refreshToken = refreshTokenService.buildRefreshToken(authentication, AuthenticationConstants.REFRESH_TOKEN_EXPIRE_IN_MINUTES);
            refreshTokenService.addToken(refreshToken);
            return new AuthenticationResponse(jwt, refreshToken.getValue());
        }
        throw new LoginException("Invalid username or password");
    }




@Service
public class UserAuthenticationService implements AuthenticationManager {

    private final UserService userService;

    @Autowired
    public UserAuthenticationService(UserService userService) {
        this.userService = userService;
    }

    @Override
    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        Optional<User> userOptional = userService.getUserByUserName(authentication.getName());
        if (userOptional.isPresent()) {
            User user = userOptional.get();
            if (user.getPassword().equals(authentication.getCredentials())) {
                return new UsernamePasswordAuthenticationToken(authentication.getPrincipal(), null, user.getAuthorities());
            }
        }
        return new UsernamePasswordAuthenticationToken(authentication.getPrincipal(), authentication.getCredentials());
    }


Dodatkowo czy Twoje hasłą nie są przechowywane w bazie danych jako plain tekst?

CezaryZal replied:
Ok, dzięki za pokazanie mechanizmu.
Niestety logikę i baze danych mam opartą na własnym obiekcie i podpięcie tego zajełoby trochę czasy dlatego narazie zostawie obecne rozwiazanie.
Napewno w przyszłości będę chciał ją wykorzystać

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/LoginFormValidator.java:
```diff
@@ -0,0 +1,39 @@
+        isToShort(login);
+    }
+
+    private void throwExceptionIfLoginIsNull(String login) {
+        if (login == null){
+            throw new NullInputException("Input login is null");
+        }
+    }
+
+    private void throwExceptionIfLoginIsEmptyLogin(String login) {
+        if (login.isEmpty()){
+            throw new EmptyObjectException("Input login is empty");
+        }
+    }
```
nie wiem czy jest sens rozbijać to na dwa przypadku zrobiłbym login == null || login.isEmpty i rzucił wtedy jednego exceptiona

CezaryZal replied:
rozbiłem żeby szczegółowo wiedzieć co się stało.
Wiem że to raczej sprawdzane jest przez frontend ale tutaj też chciałem to sprawdzać

CezaryZal committed:
Update to run by JRE

Remove in `handleValidationLoginAndPasswordExceptions` class IncorrectLoginOrPasswordSecurityException.class from @ExceptionHandler annotation;
Add repackage in goals plugin in pom.xml;
Extends `Application` class through SpringBootServletInitializer

CezaryZal committed:
Update based on comments from git pull

Remove @ResponseStatus annotation from `EmptyObjectException` class;
Rename method from `LoginService` class to throwExceptionIfUserIsNotActive;
Change return HttpStatus from `handleLoginExceptions` method to FORBIDDEN;
Move System.currentTimeMillis() to inside builder() and rename `User` object to User;
Change field in `User` entity to approved to active and database;
Create `UserNotFoundException` class and add him to method in `UserServiceImp` class;
Change name entity to `AuthenticationRequest` from UserLogin;
Change name entity to `UserToDb` from User;
Rename methods in `UserLoginValidatorService` class,
